File: AdventOfCode2021/CSharp/Day2.cs
using System.Linq;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace AdventOfCode2021
{
    public class Day2
    {
        public (int x, int y) Part1(string input)
        {
            var lines = input.Split("\r\n");
            var instructions = lines.Select(it => it.Split(" "));

            var x = 0;
            var y = 0;

            foreach (var instruction in instructions)
            {
                var magnitude = int.Parse(instruction[1]);
                switch (instruction[0])
                {
                    case "forward":
                        x += magnitude;
                        break;
                    case "up":
                        y -= magnitude;
                        break;
                    case "down":
                        y += magnitude;
                        break;
                }
            }

            return (x, y);
        }
        
        public (int x, int y) Part2(string input)
        {
            var lines = input.Split("\r\n");
            var instructions = lines.Select(it => it.Split(" "));

            var x = 0;
            var y = 0;
            var aim = 0;

            foreach (var instruction in instructions)
            {
                var magnitude = int.Parse(instruction[1]);
                switch (instruction[0])
                {
                    case "forward":
                        x += magnitude;
                        y += aim * magnitude;
                        break;
                    case "up":
                        aim -= magnitude;
                        break;
                    case "down":
                        aim += magnitude;
                        break;
                }
            }

            return (x, y);
        }
    }
    
    [TestClass]
    public class Day2Tests
    {
        private string _smallInput = @"forward 5
down 5
forward 8
up 3
down 8
forward 2";
        
        [TestMethod]
        public void TestPart1()
        {
            var expected = (15, 10);
            var actual = new Day2().Part1(_smallInput);
            Assert.AreEqual(expected, actual);
        }
        
        [TestMethod]
        public void TestPart2()
        {
            var expected = (15, 60);
            var actual = new Day2().Part2(_smallInput);
            Assert.AreEqual(expected, actual);
        }
    }
}

File: AdventOfCode2021/FSharp/Day2.fs
module FSharp.Day2

open FSharp.Core
open Microsoft.VisualStudio.TestTools.UnitTesting

type State =
    { X: int; Y: int; Aim: int }
    static member empty = { X = 0; Y = 0; Aim = 0 }

let parseInstructions (input: string) =
    input.Split("\r\n")
    |> Array.map (fun it -> it.Split(" "))

let part1 (input: string) : (int * int) =
    let instructions = input |> parseInstructions

    let folder (state: State) (current: string array) : State =
        let magnitude = current.[1] |> int

        match current.[0] with
        | "forward" -> { state with X = state.X + magnitude }
        | "up" -> { state with Y = state.Y - magnitude }
        | "down" -> { state with Y = state.Y + magnitude }
        | _ -> failwith "invalid instruction"

    instructions
    |> Array.fold folder State.empty
    |> fun it -> (it.X, it.Y)

let part2 (input: string) : (int * int) =
    let instructions = input |> parseInstructions

    let folder (state: State) (current: string array) : State =
        let magnitude = current.[1] |> int

        match current.[0] with
        | "forward" ->
            { state with
                  X = state.X + magnitude
                  Y = state.Y + (state.Aim * magnitude) }
        | "up" ->
            { state with
                  Aim = state.Aim - magnitude }
        | "down" ->
            { state with
                  Aim = state.Aim + magnitude }
        | _ -> failwith "invalid instruction"

    instructions
    |> Array.fold folder State.empty
    |> fun it -> (it.X, it.Y)

module Test =
    open Xunit

    let smallInput =
        @"forward 5
down 5
forward 8
up 3
down 8
forward 2"

    [<Fact>]
    let ``Part 1 Test`` () =
        let expected = (15, 10)
        let actual = part1 smallInput
        Assert.Equal(expected, actual)

    [<Ignore>]
    [<Fact>]
    let ``Part 2 Test`` () =
        let expected = (15, 60)
        let actual = part2 smallInput
        Assert.Equal(expected, actual)
